gbn/syncer.go

```diff
@@ -0,0 +1,291 @@
+		c.awaitedACK, c.awaitedNACK)
+}
+
+// getState returns the current state of the syncer.
+func (c *syncer) getState() syncState {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	return c.state
+}
+
+// waitForSync waits for the sync to be completed. The sync is completed when we
+// receive either the awaitedNACK, the awaitedACK + resend timeout has passed,
+// or when timing out.
+func (c *syncer) waitForSync() {
+	c.log.Tracef("Awaiting sync after resending the queue")
+
+	select {
+	case <-c.quit:
+		return
+
+	case <-c.cancel:
+		c.log.Tracef("sync canceled or reset")
+
+	case <-time.After(c.timeout * awaitingTimeoutMultiplier):
+		c.log.Tracef("Timed out while waiting for sync")
+	}
+
+	c.reset()
+}
+
+// processAck mark the sync as completed if the passed sequence number matches
+// the awaitedACK, after the resend timeout has passed.
+// If we are not resending or waiting after a resend, this is a no-op.
+func (c *syncer) processAck(seq uint8) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// If we are not resending or waiting after a resend, just swallow the
+	// ACK.
+	if c.state != syncStateResending {
+		return
+	}
+
+	// Else, if we are waiting but this is not the ack we are waiting for,
+	// just swallow it.
+	if seq != c.awaitedACK {
+		return
+	}
+
+	c.log.Tracef("Got awaited ACK")
+
+	// We start the proceedAfterTime function in a goroutine, as we
+	// don't want to block the processing of other NACKs/ACKs while
+	// we're waiting for the resend timeout to expire.
+	go c.proceedAfterTime()
+}
+
+// processNack mark the sync as completed if the passed sequence number matches
+// the awaitedNACK.
+// If we are not resending or waiting after a resend, this is a no-op.
+func (c *syncer) processNack(seq uint8) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// If we are not resending or waiting after a resend, just swallow the
+	// NACK.
+	if c.state != syncStateResending {
+		return
+	}
+
+	// Else, if we are waiting but this is not the NACK we are waiting for,
+	// just swallow it.
+	if seq != c.awaitedNACK {
+		return
+	}
+
+	c.log.Tracef("Got awaited NACK")
+
+	c.resetUnsafe()
+}
+
+// proceedAfterTime will wait for the resendTimeout and then complete the sync,
+// if we haven't completed the sync yet by receiving the awaitedNACK.
+func (c *syncer) proceedAfterTime() {
+	// We await for the duration of the resendTimeout before completing the
+	// sync, as that's the time we'd expect it to take for the other party
+	// to respond with a NACK, if the resent last packet in the
+	// queue would lead to a NACK. If we receive the awaitedNACK
+	// before the timeout, the cancel channel will be sent over, and we can
+	// stop the execution early.
+	select {
+	case <-c.quit:
+		return
+
+	case <-c.cancel:
+		c.log.Tracef("sync succeeded or was reset")
+
+		// As we can't be sure that waitForSync cancel listener was
+		// triggered before this one, we send over the cancel channel
+		// again, to make sure that both listeners are triggered.
+		c.reset()
```

Comment:
We could potentially cover this instead by closing the `cancel` channel when executing `reset` function, and creating a new channel on every sync init (and also use a `sync.Once` to ensure we only cancel the channel once per sync). But I figured this version is probably a bit more clean.

Reply:
> But then you would need to re-init the sync.Once each time too which kinda defeats the purpose right?

Yes exactly, which is why I opted for this solution :)!

> If you want to make sure that the reset only does something if we actually need it to,
then, in the reset method, you could check that we are in the "Resending" state, if we are not, it becomes a no-op. That way, the state machine becomes a bit more safe. we could add the same type of check anywhere else where we change the state

Hmmm I'm not sure how this would address that we wouldn't need to call the `c.reset()` both here and at the end of the `waitForSync` function as we can't be sure which of the select cases here + or in the `waitForSync` will catch the `<-c.cancel` case first? We need to ensure that both trigger when `c.cancel` channel get's sent over. Is this a case where you'd use a channel with a size of 2, or is there a better way to solve this?